Add Mul postprocessor for Vector

jhaayushkumar opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Related to gh-26121 and gh-26623.

#### Brief description of what is fixed or changed

When [simplify()](cci:1://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/tests/test_vector.py:193:0-225:60) traverses expression trees using `bottom_up()`, it reconstructs nodes via `.func(*args)`. This caused [VectorMul](cci:2://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/vector.py:520:0-539:35) objects to become [Mul](cci:2://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/vector.py:520:0-539:35) objects, which then failed type checks in `CoordSys3D.__new__()` with `TypeError: location should be a Vector`.

This PR adds a postprocessor for [Mul](cci:2://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/vector.py:520:0-539:35) (similar to the existing [Add](cci:2://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/vector.py:498:0-517:27) postprocessor from gh-26623) to ensure [Mul](cci:2://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/vector.py:520:0-539:35) operations on vectors return [VectorMul](cci:2://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/vector.py:520:0-539:35) objects.



#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* vector
  * Fixed TypeError when simplifying vector expressions. VectorMul objects are now preserved during simplification.

<!-- END RELEASE NOTES -->

jhaayushkumar commented:
I'm not sure about this one but the understanding that i gained from  failing of `object has no attribute '_components'. Did you mean: 'comp...`
```python
=============================== short test summary info ===============================

FAILED sympy/printing/tests/test_latex.py::test_latex_vector_expressions - AssertionError: assert '\\mathbf{\\hat{k}_{A}}' == 'x \\left(\\m..._{A}}\\right)'

FAILED sympy/printing/tests/test_mathml.py::test_print_Vector - assert '<msub><mover...A</mi></msub>' == '<mrow><mi>x<.../mrow></mrow>'

FAILED sympy/printing/pretty/tests/test_pretty.py::test_vector_expr_pretty_printing - AssertionError: assert 'k_A' == 'x⋅(i_A)×(j_A)'

FAILED sympy/vector/tests/test_field_functions.py::test_mixed_coordinates - assert Divergence(Cross(a.i, b.j)) == ((((((a.x * b.x) * c.x) * Divergence(Cross(a...

FAILED sympy/vector/tests/test_vector.py::test_cross - AttributeError: 'Cross' object has no attribute '_components'. Did you mean: 'comp...

= 5 failed, 12901 passed, 682 skipped, 329 xfailed, 2 xpassed, 8 warnings in 261.22s (0:04:21) =
```
`Cross(Vector)` made code expect `a _components` attribute that didn't exist, causing crashes.
so i changed `Cross(Vector)` to `Cross(Expr)` - now it's just an unevaluated formula (like sin(x)) not a computed vector, so no `_components` needed and everything works.

jhaayushkumar commented:
> It should just pass all the args to `VectorAdd`. The `VectorAdd` function can decide what to do with them and should raise an error if any are not vectors.
>

```diff
if vec_class == VectorAdd:
-    return VectorAdd(*vectors).doit(deep=False)
+    return VectorAdd(*expr.args).doit(deep=False)
def __new__(cls, *args, **options):
+   for arg in args:
+       if not isinstance(arg,Vector):
+           raise TypeError("VectorAdd expects all arguments to be Vector instances")
```

did it ok now?
and sorry for the delay sir, as i was busy in my semester exams.

oscarbenjamin commented on sympy/vector/vector.py:
```diff
@@ -418,19 +418,22 @@ def _div_helper(one, other):
-            return VectorAdd(*vectors).doit(deep=False)
+            return VectorAdd(*expr.args).doit(deep=False)
+        if vec_class == VectorMul:
+            return VectorMul(*vectors).doit(deep=False)
```
Can this not just pass all of `expr.args`? What happens to the other args?

I think there needs to be some tests in the `test_vector.py` for what happens when calling `Add(*args)` or `Mul(*args)` with args that are vectors or scalars. These can go along with the tests for kind.

Davide-sd commented on sympy/vector/tests/test_vector.py:
```diff
@@ -341,3 +341,73 @@ def test_scalar():
+    raises(ValueError, lambda: Mul(C.j, C.k))
+
+    result_same = Mul(C.i, C.i)
+    assert result_same == C.i**2
```
What is going on here?

oscarbenjamin replied:
I don't see why this was resolved. This behaviour does not seem correct to me.

jhaayushkumar replied:
sympy converts `Mul(C.i, C.i) to C.i**2` before our code runs. I thought since this edge case exists and skip our validation, I should add a test to track it. 
but I removed it because I realized C.i**2 doesn't make mathematical sense for vectors.

Davide-sd replied:
Maybe we can add another postprocessor for the `Pow` class. Something like this:

```py
        vec_class = {Add: VectorAdd, Mul: VectorMul, 
            # None because there is no VectorPow class
            Pow: None}
        ...
        elif vec_class is None:
            if isinstance(expr.args[0], Vector):
                raise ValueError("Power operation is not supported for vectors")

...

Basic._constructor_postprocessor_mapping[Vector] = {
    "Add": [get_postprocessor(Add)],
    "Mul": [get_postprocessor(Mul)],
    "Pow": [get_postprocessor(Pow)],
}
```

Upabjojr replied:
it's not just power... also `Mul(C.i, C.j)` should not be allowed. Vectors have either dot-product or cross-product (at least in 3D mechanics), they are both operations unrelated to standard multiplication.

jhaayushkumar committed:
Fix issue #28727: Add Mul postprocessor and vector operation validation

Add Mul to postprocessor mapping, VectorAdd type validation, and change Cross to inherit from Expr

jhaayushkumar committed:
Merge branch 'master' into fix-issue-28727

jhaayushkumar committed:
remove trailing space

jhaayushkumar committed:
add space

oscarbenjamin commented:
The outstanding issue here is:
```
In [1]: from sympy.vector import CoordSys3D, express

In [2]: C = CoordSys3D('C')

In [3]: Mul(C.i, C.i)
Out[3]: 
   2
i_C 
```

jhaayushkumar committed:
Add Pow postprocessor to reject vector exponentiation

Fixes Mul(C.i, C.i) bypass converting to C.i**2. Adds comprehensive tests covering all vector types and exponent cases.

jhaayushkumar commented:
Fixed! the Pow postprocessor catches Mul(C.i, C.i) and raises ValueError. also add test for C.i**2 .

jhaayushkumar committed:
Change error type to TypeError per Oscar's feedback

jhaayushkumar committed:
Merge branch 'master' into fix-issue-28727

oscarbenjamin commented on sympy/vector/vector.py:
```diff
@@ -418,19 +418,22 @@ def _div_helper(one, other):
     "Add": [get_postprocessor(Add)],
+    "Mul": [get_postprocessor(Mul)],
+    "Pow": [get_postprocessor(Pow)],
 }
```
The `get_postprocessor` function seems like a pointless complication. Better just to have
```
def _add_vector_processor(expr: Add) -> Vector:
    return VectorAdd(*expr.args).doit(deep=False)
```

jhaayushkumar committed:
Refactor vector postprocessors to separate functions

jhaayushkumar committed:
Simplify Pow postprocessor to always raise TypeError

jhaayushkumar committed:
Add tests for exponent as vector cases

oscarbenjamin commented:
I think this looks good now. @Upabjojr or @Davide-sd do you want to review?

jhaayushkumar commented on sympy/vector/vector.py:
```diff
@@ -499,6 +498,9 @@ class VectorAdd(BasisDependentAdd, Vector):
     def __new__(cls, *args, **options):
+        for arg in args:
+            if not isinstance(arg, Vector):
+                raise TypeError("VectorAdd expects all arguments to be Vector instances")
```
I thought an explicit check would be safer and provide a much clearer error message than the default one.
But agreed, it's redundant, so should i remove it ?

jhaayushkumar replied:
remove the redundant part.

jhaayushkumar replied:
@Davide-sd did it ok now?

jhaayushkumar committed:
Remove redundant validation in VectorAdd

oscarbenjamin commented:
Okay thanks all. Let''s get it in.
